File: app/src/main/java/com/ericho/coupleshare/mvp/data/remote/LoginRemoteDataSource.java
package com.ericho.coupleshare.mvp.data.remote;

import android.content.Context;
import android.os.Handler;
import android.os.HandlerThread;

import com.ericho.coupleshare.http.BaseUiCallback;
import com.ericho.coupleshare.http.GsonUtil;
import com.ericho.coupleshare.http.model.BaseSingleResponse;
import com.ericho.coupleshare.mvp.data.LoginDataSource;
import com.ericho.coupleshare.util.NetworkUtil;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.net.SocketTimeoutException;

import okhttp3.Call;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import timber.log.Timber;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by steve_000 on 12/6/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.mvp.data.remote
 */

public class LoginRemoteDataSource implements LoginDataSource {

    private static LoginRemoteDataSource INSTANCE ;

    private HandlerThread mHandlerThread;

    private Handler mHandler;

    private OkHttpClient client;

    private Context context;

    private LoginRemoteDataSource(Context context){
        this.context = context;
        mHandlerThread = new HandlerThread("user-remote");
        mHandlerThread.start();
        mHandler = new Handler(mHandlerThread.getLooper());
        client = NetworkUtil.getOkhttpClient();
    }

    public static LoginRemoteDataSource getInstance(Context context){
        if(INSTANCE == null){
            INSTANCE = new LoginRemoteDataSource(context);
        }
        return INSTANCE;
    }

    @Override
    public void login(String username, String password, LoginCallback callback) {
        mHandler.post(() -> {
            try {
                Request request = NetworkUtil.loginAccount(context,username,password);
                Call call = NetworkUtil.execute(client,request);
                Response response = call.execute();
                String res = NetworkUtil.parseResponseToString(response);
                Timber.d(res);
                Type type = new TypeToken<BaseSingleResponse<String>>(){}.getType();
                BaseSingleResponse<String> baseSingleResponse = GsonUtil.getGson().fromJson(res,type);
                if (baseSingleResponse.getStatus()) {
                    callback.onLoginSuccess();
                }else {
                    callback.onLoginFailure(new IOException(baseSingleResponse.getErrorMessage()));
                }
            } catch (SocketTimeoutException e) {
                callback.onLoginFailure(e);
            }catch (IOException e) {
                e.printStackTrace();
                callback.onLoginFailure(e);
            }
        });


    }

    @Override
    public void register(String username, String password, RegisterCallback callback) {
        mHandler.post(() -> {
            try {
                Request request = NetworkUtil.registerAccount(context,username,password);
                Call call = NetworkUtil.execute(client,request);
                Response response = call.execute();
                String res = NetworkUtil.parseResponseToString(response);

                Type type = new TypeToken<BaseSingleResponse<String>>(){}.getType();
                BaseSingleResponse<String> baseSingleResponse = GsonUtil.getGson().fromJson(res,type);
                if (baseSingleResponse.getStatus()) {
                    callback.onRegisterSuccess();
                }else {
                    callback.onRegisterFailure(new IOException(baseSingleResponse.getErrorMessage()));
                }
            } catch (IOException e) {
                e.printStackTrace();
                callback.onRegisterFailure(e);
            }
        });
    }
}

File: app/src/main/java/com/ericho/coupleshare/mvp/data/remote/LoginRemoteDataSource.kt
package com.ericho.coupleshare.mvp.data.remote

import android.content.Context
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import com.ericho.coupleshare.App
import com.ericho.coupleshare.network.model.BaseSingleResponse
import com.ericho.coupleshare.mvp.data.LoginDataSource
import com.ericho.coupleshare.util.NetworkUtil
import com.google.gson.reflect.TypeToken
import okhttp3.OkHttpClient
import timber.log.Timber
import java.io.IOException
import java.net.SocketTimeoutException

/**
 * Created by steve_000 on 8/7/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.mvp.data.remote
 */
class LoginRemoteDataSource private constructor(var context: Context):LoginDataSource {


    private val mHandlerThread: HandlerThread

    private val mHandler: Handler
    private val mMainHandler: Handler

    private val client: OkHttpClient

    init {
        mHandlerThread = HandlerThread("user-remote")
        mHandlerThread.start()
        mHandler = Handler(mHandlerThread.getLooper())
        mMainHandler = Handler(Looper.getMainLooper())
        client = NetworkUtil.getOkhttpClient()
    }

    override fun login(username: String, password: String, callback: LoginDataSource.LoginCallback) {
        mHandler.post {
            try {
                val request = NetworkUtil.loginAccount(context, username, password)
                val call = NetworkUtil.execute(client, request)
                val response = call.execute()
                val res = NetworkUtil.parseResponseToString(response)
                Timber.d(res)
                val type = object : TypeToken<BaseSingleResponse<String>>() {}.type
                val baseSingleResponse = App.gson.fromJson<BaseSingleResponse<String>>(res, type)
                if (baseSingleResponse.status) {
                    callback.onLoginSuccess()
                } else {
                    callback.onLoginFailure(IOException(baseSingleResponse.errorMessage))
                }
            } catch (e: SocketTimeoutException) {
                callback.onLoginFailure(e)
            } catch (e: IOException) {
                e.printStackTrace()
                callback.onLoginFailure(e)
            }
        }
    }

    override fun register(username: String, password: String, callback: LoginDataSource.RegisterCallback) {
        mHandler.post {
            try {
                val request = NetworkUtil.registerAccount(context, username, password)
                val call = NetworkUtil.execute(client, request)
                val response = call.execute()
                val res = NetworkUtil.parseResponseToString(response)

                val type = object : TypeToken<BaseSingleResponse<String>>() {}.type
                val baseSingleResponse = App.gson.fromJson<BaseSingleResponse<String>>(res, type)
                if (baseSingleResponse.status) {
                    callback.onRegisterSuccess()
                } else {
                    callback.onRegisterFailure(IOException(baseSingleResponse.errorMessage))
                }
            } catch (e: IOException) {
                e.printStackTrace()
                callback.onRegisterFailure(e)
            }
        }
    }

    companion object {
        private var INSTANCE : LoginRemoteDataSource? = null

        @JvmStatic
        fun getInstance(context: Context):LoginRemoteDataSource{
            if(INSTANCE==null){
                INSTANCE = LoginRemoteDataSource(context)
            }
            return INSTANCE as LoginRemoteDataSource
        }
    }
}